### PR TITLE
Overhaul of query field treatment in associations: instead of keeping…

### PR DESCRIPTION
**Pull request by bserdar**
… absolute field names, we now keep field names with their entities, and rewrite queries relative to a node. What used to work still works, what was broken is still broken, but at least now I can fix the broken stuff, namely, reference.*.field=x search

### CODE REVIEW

File: crud/src/test/java/com/redhat/lightblue/assoc/RelativePathTest.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+package com.redhat.lightblue.assoc;
+
+import org.junit.Test;
+import org.junit.Assert;
+
+import com.redhat.lightblue.util.Path;
+
+public class RelativePathTest {
+    @Test
+    public void testRelativePaths() throws Exception {
```
Comment (jewzaam) [inline crud/src/test/java/com/redhat/lightblue/assoc/RelativePathTest.java]: Would like to see path have field different from fullPath.
And add tests with cmdPrefix non-empty.

File: query-api/src/main/java/com/redhat/lightblue/query/MapQueryFieldsIterator.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+     * Override this method to map fields.
+     */
+    protected Path map(Path p) {
+        return p;
+    }
+    
+    @Override
+    protected QueryExpression itrValueComparisonExpression(ValueComparisonExpression q, Path context) {
+        Path p=map(applyContext(context,q.getField()));
+        if(p!=null)
```
Comment (jewzaam) [inline query-api/src/main/java/com/redhat/lightblue/query/MapQueryFieldsIterator.java]: With the default implementation a new expression is always created.  Could it be updated to only do that when p!=q.field?

File: crud/src/test/java/com/redhat/lightblue/assoc/RelativePathTest.java (hunk 1 above)
Comment (bserdar) [inline crud/src/test/java/com/redhat/lightblue/assoc/RelativePathTest.java]: Added test cases where cmdPrefix is nonempty.

Added some comments to the function itself. The path field cannot be
different from fullPath for this function, it gets the fullpath to field,
the field as it appears in the expression, and the entity prefix containing
the field, and writes an entity relative field name, so a suffix of the
field name has to match a suffix of the fullpath.

File: query-api/src/main/java/com/redhat/lightblue/query/MapQueryFieldsIterator.java (hunk 2 above)
Comment (bserdar) [inline query-api/src/main/java/com/redhat/lightblue/query/MapQueryFieldsIterator.java]: Added comments. The map function should return null if the field is not to
be mapped.
